builtin/docker/pull/kaniko.go

```diff
@@ -0,0 +1,158 @@
+			oci.AuthConfig.Username = user
+			oci.AuthConfig.Password = pass
+		case *wpdocker.AccessInfo_Header:
+			oci.AuthConfig.Auth = sv.Header
+		case *wpdocker.AccessInfo_UserPass_:
+			oci.AuthConfig.Username = sv.UserPass.Username
+			oci.AuthConfig.Password = sv.UserPass.Password
+		}
+	}
+
+	// Determine the host that we're setting auth for. We have to parse the
+	// image for this cause it may not contain a host. Luckily Docker has
+	// libs to normalize this all for us.
+	log.Trace("determining host for auth configuration", "image", target.Name())
+	ref, err := reference.ParseNormalizedNamed(target.Image)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to parse image name: %s", err)
+	}
+
+	host := reference.Domain(ref)
+	if host == "docker.io" {
+		// The normalized name parse above will turn short names like "foo/bar"
+		// into "docker.io/foo/bar" but the actual registry host for these
+		// is "index.docker.io".
+		host = "index.docker.io"
+	}
+	log.Trace("auth host", "host", host)
+
+	if ai.Insecure {
+		oci.Upstream = "http://" + host
+	} else {
+		oci.Upstream = "https://" + host
+	}
+
+	refPath := reference.Path(ref)
+
+	if !b.config.DisableCEB {
+		// For Kaniko we can use our runtime arch because the image we build
+		// always matches the architecture of our Kaniko environment.
+		assetName, ok := assets.CEBArch[runtime.GOARCH]
+		if !ok {
+			return nil, status.Errorf(codes.FailedPrecondition,
+				"automatic entrypoint injection not supported for architecture: %s", runtime.GOARCH)
+		}
+
+		data, err := assets.Asset(assetName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "unable to restore custom entry point binary: %s", err)
+		}
+
+		step.Done()
+		step = sg.Add("Testing registry and uploading entrypoint layer")
+
+		err = oci.SetupEntrypointLayer(refPath, data)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error setting up entrypoint layer: %s", err)
+		}
+	}
+
+	li, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, err
+	}
+
+	defer li.Close()
+	go http.Serve(li, &oci)
+
+	port := li.Addr().(*net.TCPAddr).Port
+
+	localRef := fmt.Sprintf("localhost:%d/%s:%s", port, refPath, ai.Tag)
+
+	dockerfileBS := []byte(fmt.Sprintf("FROM %s:%s\n", target.Image, target.Tag))
+	err = os.WriteFile("Dockerfile", dockerfileBS, 0644)
+	if err != nil {
+		return nil, err
+	}
+	// Start constructing our arg string for img
+	args := []string{
+		"/kaniko/executor",
+		"-f", filepath.Dir("Dockerfile"),
+		"-d", localRef,
+	}
+
+	log.Debug("executing kaniko", "args", args)
+
+	step.Done()
+	step = sg.Add("Executing kaniko...")
+
+	// Command output should go to the step
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
+	cmd.Stdout = step.TermOutput()
+	cmd.Stderr = cmd.Stdout
+
+	// check for error from executor run
+	if err := cmd.Run(); err != nil {
+		return nil, err
+	}
+
+	step.Done()
+
+	step = sg.Add("Image pushed to '%s:%s'", ai.Image, ai.Tag)
```

Comment:
I might be a little confused with the UX messages, but this is pulling with Kaniko, but the final message is pushed? I think we're pushing because we inject an entrypoint after pulling the original container? We could use some clearer messaging at the end here.

Reply:
@briancain I will definitely try to clear this up. This actually stumped me for a bit when I was reading through the Docker builder code, but I think it's because Kaniko is doing a Docker build, tag, and push all-in-one, even if that push is to a local registry.